PSA-eBPF parser metadata issue

The [new PSA-eBPF backend](https://github.com/p4lang/p4c/tree/main/backends/ebpf/psa) doesn't support parser metadata correctly.

When I try to use the metadata with the type of `psa_ingress_parser_input_metadata_t`, I get an error message from the clang compiler.

Input:
```
parser packet_parser(...,
                     in psa_ingress_parser_input_metadata_t pipi_md,
                     ...) {

    ...
    
    state parse_ethernet{
        pkt.extract(hdr.ethernet);
        transition select(pipi_md.packet_path){
            PSA_PacketPath_t.RESUBMIT: reject;
            default: accept;
        }
    }

}
```

Error message:
```
out.c:341:20: error: use of undeclared identifier 'pipi_md'
        select_0 = pipi_md.packet_path;
                   ^
```

For reproducibility purposes, I created this [repository with a minimal example](https://github.com/gycsaba96/psa-ebpf-parser-metadata-issue). One can reproduce the issue using the following commands:

```
make p4c-docker-setup
make build
```

Looking at the generated C code, the metadata is defined after the `accept` label with the wrong name:

```
accept: {
    struct psa_ingress_input_metadata_t standard_metadata = {
        .ingress_port = skb->ifindex,
        .packet_path = compiler_meta__->packet_path,
        .parser_error = ebpf_errorCode,
};
```

--- a/backends/ebpf/ebpfParser.h
+++ b/backends/ebpf/ebpfParser.h
@@ -92,7 +92,7 @@ class EBPFParser : public EBPFObject {
     explicit EBPFParser(const EBPFProgram* program, const IR::ParserBlock* block,
                         const P4::TypeMap* typeMap);
     virtual void emitDeclaration(CodeBuilder* builder, const IR::Declaration* decl);
-    void emit(CodeBuilder* builder);
+    virtual void emit(CodeBuilder* builder);
     virtual bool build();
 
     virtual void emitTypes(CodeBuilder* builder);

--- a/backends/ebpf/psa/ebpfPipeline.cpp
+++ b/backends/ebpf/psa/ebpfPipeline.cpp
@@ -87,6 +87,11 @@ void EBPFPipeline::emitLocalVariables(CodeBuilder* builder) {
         emitTimestamp(builder);
         builder->endOfStatement(true);
     }
+
+    builder->emitIndent();
+    builder->appendFormat("u32 %s = %s;", inputPortVar.c_str(), ifindexVar.c_str());
+    builder->newline();
+    emitInputPortMapping(builder);
 }
 
 void EBPFPipeline::emitUserMetadataInstance(CodeBuilder* builder) {
@@ -169,6 +174,17 @@ void EBPFPipeline::emitTimestamp(CodeBuilder* builder) {
     builder->appendFormat("bpf_ktime_get_ns()");
 }
 
+void EBPFPipeline::emitInputPortMapping(CodeBuilder* builder) {
+    builder->emitIndent();
+    builder->appendFormat("if (%s == PSA_PORT_RECIRCULATE) ", inputPortVar.c_str());
+    builder->blockStart();
+    builder->emitIndent();
+    // To be conformant with psa.p4, where PSA_PORT_RECIRCULATE is constant
+    builder->appendFormat("%s = P4C_PSA_PORT_RECIRCULATE", inputPortVar.c_str());
+    builder->endOfStatement(true);
+    builder->blockEnd(true);
+}
+
 // =====================EBPFIngressPipeline===========================
 void EBPFIngressPipeline::emitSharedMetadataInitializer(CodeBuilder* builder) {
     auto type = EBPFTypeFactory::instance->create(this->deparser->resubmit_meta->type);
@@ -184,7 +200,7 @@ void EBPFIngressPipeline::emitPSAControlInputMetadata(CodeBuilder* builder) {
         "            .packet_path = %s,\n"
         "            .parser_error = %s,\n"
         "    };",
-        control->inputStandardMetadata->name.name, ifindexVar.c_str(), packetPathVar.c_str(),
+        control->inputStandardMetadata->name.name, inputPortVar.c_str(), packetPathVar.c_str(),
         errorVar.c_str());
     builder->newline();
     if (shouldEmitTimestamp()) {
@@ -247,7 +263,7 @@ void EBPFIngressPipeline::emit(CodeBuilder* builder) {
         "pkt_len=%%d",
         sectionName);
     varStr = Util::printf_format("%s->packet_path", compilerGlobalMetadata);
-    builder->target->emitTraceMessage(builder, msgStr.c_str(), 3, ifindexVar.c_str(), varStr,
+    builder->target->emitTraceMessage(builder, msgStr.c_str(), 3, inputPortVar.c_str(), varStr,
                                       lengthVar.c_str());
 
     // PARSER
@@ -353,7 +369,7 @@ void EBPFEgressPipeline::emitPSAControlInputMetadata(CodeBuilder* builder) {
         "            .instance = %s,\n"
         "            .parser_error = %s,\n"
         "        };",
-        control->inputStandardMetadata->name.name, priorityVar.c_str(), ifindexVar.c_str(),
+        control->inputStandardMetadata->name.name, priorityVar.c_str(), inputPortVar.c_str(),
         packetPathVar.c_str(), pktInstanceVar.c_str(), errorVar.c_str());
     builder->newline();
     if (shouldEmitTimestamp()) {
@@ -362,16 +378,6 @@ void EBPFEgressPipeline::emitPSAControlInputMetadata(CodeBuilder* builder) {
                               timestampVar.c_str());
         builder->endOfStatement(true);
     }
-    builder->emitIndent();
-    builder->appendFormat("if (%s.egress_port == PSA_PORT_RECIRCULATE) ",
-                          control->inputStandardMetadata->name.name);
-    builder->blockStart();
-    builder->emitIndent();
-    // To be conformant with psa.p4, where PSA_PORT_RECIRCULATE is constant
-    builder->appendFormat("%s.egress_port = P4C_PSA_PORT_RECIRCULATE",
-                          control->inputStandardMetadata->name.name);
-    builder->endOfStatement(true);
-    builder->blockEnd(true);
 }
 
 void EBPFEgressPipeline::emitPSAControlOutputMetadata(CodeBuilder* builder) {
@@ -427,7 +433,7 @@ void EBPFEgressPipeline::emit(CodeBuilder* builder) {
         "pkt_len=%%d",
         sectionName);
     varStr = Util::printf_format("%s->packet_path", compilerGlobalMetadata);
-    builder->target->emitTraceMessage(builder, msgStr.c_str(), 3, ifindexVar.c_str(), varStr,
+    builder->target->emitTraceMessage(builder, msgStr.c_str(), 3, inputPortVar.c_str(), varStr,
                                       lengthVar.c_str());
 
     // PARSER

--- a/backends/ebpf/psa/ebpfPipeline.h
+++ b/backends/ebpf/psa/ebpfPipeline.h
@@ -49,6 +49,8 @@ class EBPFPipeline : public EBPFProgram {
     cstring oneKey;
     // A unique mark used to differentiate packets processed by P4/eBPF from others.
     unsigned packetMark;
+    // A variable to store ifindex after mapping (e.g. due to recirculation)
+    cstring inputPortVar;
 
     EBPFControlPSA* control;
     EBPFDeparserPSA* deparser;
@@ -57,6 +59,7 @@ class EBPFPipeline : public EBPFProgram {
                  P4::TypeMap* typeMap)
         : EBPFProgram(options, nullptr, refMap, typeMap, nullptr),
           name(name),
+          packetMark(0x99),
           control(nullptr),
           deparser(nullptr) {
         sectionName = "classifier/" + name;
@@ -73,7 +76,7 @@ class EBPFPipeline : public EBPFProgram {
         pktInstanceVar = compilerGlobalMetadata + cstring("->instance");
         priorityVar = cstring("skb->priority");
         oneKey = EBPFModel::reserved("one");
-        packetMark = 0x99;
+        inputPortVar = cstring("ebpf_input_port");
         progTarget = new KernelSamplesTarget(options.emitTraceMessages);
     }
 
@@ -124,6 +127,7 @@ class EBPFPipeline : public EBPFProgram {
     virtual void emitGlobalMetadataInitializer(CodeBuilder* builder);
     virtual void emitPacketLength(CodeBuilder* builder);
     virtual void emitTimestamp(CodeBuilder* builder);
+    void emitInputPortMapping(CodeBuilder* builder);
 
     void emitHeadersFromCPUMAP(CodeBuilder* builder);
     void emitMetadataFromCPUMAP(CodeBuilder* builder);

--- a/backends/ebpf/psa/ebpfPsaGen.cpp
+++ b/backends/ebpf/psa/ebpfPsaGen.cpp
@@ -794,7 +794,10 @@ bool ConvertToEBPFParserPSA::preorder(const IR::ParserBlock* prsr) {
     parser->headers = *it;
     ++it;
     parser->user_metadata = *it;
-    auto resubmit_meta = *(it + 2);
+    ++it;
+    parser->inputMetadata = *it;
+    ++it;
+    auto resubmit_meta = *it;
 
     for (auto state : prsr->container->states) {
         auto ps = new EBPFParserState(state, parser);

--- a/backends/ebpf/psa/ebpfPsaParser.cpp
+++ b/backends/ebpf/psa/ebpfPsaParser.cpp
@@ -17,6 +17,7 @@ limitations under the License.
 #include "ebpfPsaParser.h"
 
 #include "backends/ebpf/ebpfType.h"
+#include "backends/ebpf/psa/ebpfPipeline.h"
 #include "frontends/p4/enumInstance.h"
 
 namespace EBPF {

--- a/backends/ebpf/psa/ebpfPsaParser.cpp
+++ b/backends/ebpf/psa/ebpfPsaParser.cpp
@@ -37,8 +37,45 @@ void PsaStateTranslationVisitor::processMethod(const P4::ExternMethod* ext) {
 // =====================EBPFPsaParser=============================
 EBPFPsaParser::EBPFPsaParser(const EBPFProgram* program, const IR::ParserBlock* block,
                              const P4::TypeMap* typeMap)
-    : EBPFParser(program, block, typeMap) {
+    : EBPFParser(program, block, typeMap), inputMetadata(nullptr) {
     visitor = new PsaStateTranslationVisitor(program->refMap, program->typeMap, this);
+}
+
+void EBPFPsaParser::emit(CodeBuilder* builder) {
+    builder->emitIndent();
+    builder->blockStart();
+    emitParserInputMetadata(builder);
+    EBPFParser::emit(builder);
+    builder->blockEnd(true);
+}
+
+void EBPFPsaParser::emitParserInputMetadata(CodeBuilder* builder) {
+    bool isIngress = program->is<EBPFIngressPipeline>();
+
+    builder->emitIndent();
+    if (isIngress) {
+        builder->append("struct psa_ingress_parser_input_metadata_t ");
+    } else {
+        builder->append("struct psa_egress_parser_input_metadata_t ");
+    }
+    builder->append(inputMetadata->name.name);
+    builder->append(" = ");
+    builder->blockStart();
+
+    builder->emitIndent();
+    if (isIngress) {
+        builder->append(".ingress_port = ");
+    } else {
+        builder->append(".egress_port = ");
+    }
+    builder->append(program->to<EBPFPipeline>()->inputPortVar);
+    builder->appendLine(",");
+
+    builder->emitIndent();
+    builder->appendLine(".packet_path = compiler_meta__->packet_path,");
+
+    builder->blockEnd(false);
+    builder->endOfStatement(true);
 }
 
 void EBPFPsaParser::emitDeclaration(CodeBuilder* builder, const IR::Declaration* decl) {

--- a/backends/ebpf/psa/ebpfPsaParser.h
+++ b/backends/ebpf/psa/ebpfPsaParser.h
@@ -40,10 +40,13 @@ class PsaStateTranslationVisitor : public StateTranslationVisitor {
 class EBPFPsaParser : public EBPFParser {
  public:
     std::map<cstring, EBPFChecksumPSA*> checksums;
+    const IR::Parameter* inputMetadata;
 
     EBPFPsaParser(const EBPFProgram* program, const IR::ParserBlock* block,
                   const P4::TypeMap* typeMap);
 
+    void emit(CodeBuilder* builder) override;
+    void emitParserInputMetadata(CodeBuilder* builder);
     void emitDeclaration(CodeBuilder* builder, const IR::Declaration* decl) override;
     void emitRejectState(CodeBuilder* builder) override;
 

